data-prepper-plugins/s3-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/accumulator/InMemoryBuffer.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.dataprepper.plugins.sink.accumulator;
+
+import org.apache.commons.lang3.time.StopWatch;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import software.amazon.awssdk.awscore.exception.AwsServiceException;
+import software.amazon.awssdk.core.exception.SdkClientException;
+import software.amazon.awssdk.core.sync.RequestBody;
+import software.amazon.awssdk.services.s3.S3Client;
+import software.amazon.awssdk.services.s3.model.PutObjectRequest;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * A buffer can hold in memory data and flushing it to S3.
+ */
+public class InMemoryBuffer implements Buffer {
+
+    private static final Logger LOG = LoggerFactory.getLogger(InMemoryBuffer.class);
+    private static final ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
+    private int eventCount;
+    private final StopWatch watch;
+
+    InMemoryBuffer() {
+        byteArrayOutputStream.reset();
+        eventCount = 0;
+        watch = new StopWatch();
+        watch.start();
+    }
+
+    @Override
+    public long getSize() {
+        return byteArrayOutputStream.size();
+    }
+
+    @Override
+    public int getEventCount() {
+        return eventCount;
+    }
+
+    public long getDuration() {
+        return watch.getTime(TimeUnit.SECONDS);
+    }
+
+    /**
+     * Upload accumulated data to amazon s3
+     *
+     * @param s3Client s3 client object.
+     * @param bucket   bucket name.
+     * @param key      s3 object key path.
+     * @return boolean based on file upload status.
+     */
+    @Override
+    public boolean flushToS3(S3Client s3Client, String bucket, String key, final int maxRetries)
+            throws InterruptedException {
+        boolean isUploadedToS3 = Boolean.FALSE;
+        final byte[] byteArray = byteArrayOutputStream.toByteArray();
+        int retryCount = maxRetries;
+        do {
```

Comment:
We should consider moving this retry logic into the `S3SinkService` class. You are going to end up duplicating it between the InMemoryBuffer and the FileBuffer implementations. But, you can make that refactoring the follow-on PR which adds the file buffer.

Reply:
Resolved.